src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala

```diff
@@ -54,24 +55,40 @@ class CassandraJournal extends AsyncWriteJournal with CassandraRecovery with Cas
   }
 
   private def statementGroup(atomicWrite: AtomicWrite): Try[Seq[BoundStatement]] = Try {
-    atomicWrite.payload.flatMap { m =>
-      val pnr = partitionNr(m.sequenceNr)
-      val write = preparedWriteMessage.bind(m.persistenceId, pnr: JLong, m.sequenceNr: JLong, persistentToByteBuffer(m))
-      if (partitionNew(m.sequenceNr)) Seq(preparedWriteHeader.bind(m.persistenceId, pnr: JLong), write) else Seq(write)
+    // hoping to remove this in 2.4-M3 https://github.com/akka/akka/issues/18076
+    val maxPnr = partitionNr(atomicWrite.payload.last.sequenceNr)
+    val firstSeq: JLong = atomicWrite.payload.head.sequenceNr
+    val minPnr: JLong = partitionNr(firstSeq)
+    val persistenceId: String = atomicWrite.payload.head.persistenceId
+
+    // reading assumes sequence numbers are in the right partition or partition + 1
+    // even if we did allow this it would perform terribly as large C* batches are not good
+    if (maxPnr - minPnr > 1) throw new RuntimeException("Do not support AtomicWrites that span 3 partitions. Keep AtomicWrites <= max partition size.")
+
+    val writes: Seq[BoundStatement] = atomicWrite.payload.map { m =>
+      preparedWriteMessage.bind(m.persistenceId, maxPnr: JLong, m.sequenceNr: JLong, persistentToByteBuffer(m))
     }
+    // in case we skip an entire partition we want to make sure the empty partition has in in-use flag so scans
+    // keep going when they encounter it
+    if (partitionNew(firstSeq) && minPnr != maxPnr) writes :+ preparedWriteInUse.bind(persistenceId, minPnr)
+    else writes
+
   }
 
   private def asyncDeleteMessages(messageIds: Seq[MessageId]): Future[Unit] = executeBatch { batch =>
     messageIds.foreach { mid =>
-      val stmt =
-        preparedDeletePermanent.bind(mid.persistenceId, partitionNr(mid.sequenceNr): JLong, mid.sequenceNr: JLong)
+      val firstPnr: JLong = partitionNr(mid.sequenceNr)
+      val stmt = preparedDeletePermanent.bind(mid.persistenceId, firstPnr: JLong, mid.sequenceNr: JLong)
+      // the message could be in next partition as a result of an AtomicWrite, alternative is a read before write
```

Comment:
Would a read-before-delete be done for every message or can this be implemented more efficiently? If we could implement that with a single read per partition (or n reads where n is a small constant) I'd prefer going for that approach. We only append data, so race conditions shouldn't be an issue. If we need a read for every message, we should go for the redundant-delete approach (as users might delete millions of old messages with a single request).